File: ios/Core/Data/Sources/CoreData/DefaultMusicLibrary.swift
import CoreDomain
import Foundation

public final class DefaultMusicLibrary: MusicLibrary, @unchecked Sendable {
    private let providersById: [MusicProviderId: MusicProvider]
    private let orderedProviders: [MusicProvider]

    public init(providers: Set<AnyMusicProvider>) {
        let baseProviders = providers.map(\.base)
        providersById = Dictionary(uniqueKeysWithValues: baseProviders.map { ($0.id, $0) })
        orderedProviders = baseProviders.sorted { lhs, rhs in
            providerRank(lhs.id) < providerRank(rhs.id)
        }
    }

    public convenience init(providers: [MusicProvider]) {
        self.init(providers: Set(providers.map(AnyMusicProvider.init)))
    }

    public func observeAvailability() -> AsyncStream<MusicServiceAvailability> {
        guard !orderedProviders.isEmpty else {
            return stream(
                once: MusicServiceAvailability(isAvailable: false, region: nil, permissions: [])
            )
        }

        return combineLatest(orderedProviders.map { provider in
            { provider.observeAvailability() }
        }) { values in
            values.reduce(
                into: MusicServiceAvailability(isAvailable: false, region: nil, permissions: [])
            ) { partial, next in
                partial = MusicServiceAvailability(
                    isAvailable: partial.isAvailable || next.isAvailable,
                    region: partial.region ?? next.region,
                    permissions: partial.permissions.union(next.permissions)
                )
            }
        }
    }

    public func observeAllPlaylists() -> AsyncStream<[PlaylistSummary]> {
        guard !orderedProviders.isEmpty else {
            return stream(once: [])
        }

        return combineLatest(orderedProviders.map { provider in
            { provider.observePlaylists() }
        }) { playlists in
            playlists
                .flatMap { $0 }
                .sorted {
                    let lhsRank = providerRank($0.provider)
                    let rhsRank = providerRank($1.provider)
                    if lhsRank != rhsRank {
                        return lhsRank < rhsRank
                    }
                    return $0.title.localizedCaseInsensitiveCompare($1.title) == .orderedAscending
                }
        }
    }

    public func observePlaylist(ref: PlaylistRef) -> AsyncStream<Playlist?> {
        provider(for: ref.provider).observePlaylist(id: ref.id)
    }

    public func observeSavedTracks() -> AsyncStream<SavedTracksResult> {
        let savedTracksProviders = orderedProviders.filter { $0.id == .yandexMusic }
        guard !savedTracksProviders.isEmpty else {
            return stream(once: .privateLibrary)
        }

        return combineLatest(savedTracksProviders.map { provider in
            { provider.observeSavedTracks() }
        }) { results in
            results.first(where: { result in
                if case .privateLibrary = result {
                    return false
                }
                return true
            }) ?? .privateLibrary
        }
    }

    public func refreshAll() async throws {
        try await withThrowingTaskGroup(of: Void.self) { group in
            for provider in orderedProviders {
                group.addTask {
                    try await provider.refreshAvailability()
                    try await provider.refreshPlaylists()
                }
            }
            try await group.waitForAll()
        }
    }

    public func refreshPlaylist(ref: PlaylistRef) async throws {
        try await provider(for: ref.provider).refreshPlaylist(id: ref.id)
    }

    public func refreshSavedTracks() async throws {
        try await withThrowingTaskGroup(of: Void.self) { group in
            for provider in orderedProviders {
                group.addTask {
                    try await provider.refreshSavedTracks()
                }
            }
            try await group.waitForAll()
        }
    }

    private func provider(for id: MusicProviderId) -> MusicProvider {
        guard let provider = providersById[id] else {
            fatalError("Music provider \(id) is not registered.")
        }
        return provider
    }
}

public struct AnyMusicProvider: Hashable, Sendable {
    public let base: MusicProvider
    private let id = UUID()

    public init(_ base: MusicProvider) {
        self.base = base
    }

    public static func == (lhs: Self, rhs: Self) -> Bool {
        lhs.id == rhs.id
    }

    public func hash(into hasher: inout Hasher) {
        hasher.combine(id)
    }
}

private func providerRank(_ providerId: MusicProviderId) -> Int {
    switch providerId {
    case .device:
        return 0

    case .yandexMusic:
        return 1
    }
}

File: android/core/data/src/main/java/com/mplayeraudio/core/data/DefaultMusicLibrary.kt
package com.mplayeraudio.core.data

import com.mplayeraudio.core.domain.musiclibrary.MusicLibrary
import com.mplayeraudio.core.domain.musiclibrary.MusicProvider
import com.mplayeraudio.core.domain.musiclibrary.MusicProviderId
import com.mplayeraudio.core.domain.musiclibrary.MusicServiceAvailability
import com.mplayeraudio.core.domain.musiclibrary.Playlist
import com.mplayeraudio.core.domain.musiclibrary.PlaylistRef
import com.mplayeraudio.core.domain.musiclibrary.PlaylistSummary
import com.mplayeraudio.core.domain.musiclibrary.SavedTracksResult
import kotlinx.coroutines.async
import kotlinx.coroutines.awaitAll
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.combine
import kotlinx.coroutines.flow.flowOf

class DefaultMusicLibrary(
    providers: Set<MusicProvider>,
) : MusicLibrary {

    private val providersById = providers.associateBy(MusicProvider::id)
    private val orderedProviders = providers.sortedBy(MusicProvider::id)

    override fun observeAvailability(): Flow<MusicServiceAvailability> {
        if (orderedProviders.isEmpty()) {
            return flowOf(
                MusicServiceAvailability(
                    isAvailable = false,
                    region = null,
                    permissions = emptySet(),
                ),
            )
        }

        return combine(orderedProviders.map(MusicProvider::observeAvailability)) { availability ->
            availability.fold(
                MusicServiceAvailability(
                    isAvailable = false,
                    region = null,
                    permissions = emptySet(),
                ),
            ) { acc, next ->
                MusicServiceAvailability(
                    isAvailable = acc.isAvailable || next.isAvailable,
                    region = acc.region ?: next.region,
                    permissions = acc.permissions + next.permissions,
                )
            }
        }
    }

    override fun observeAllPlaylists(): Flow<List<PlaylistSummary>> {
        if (orderedProviders.isEmpty()) {
            return flowOf(emptyList())
        }

        return combine(orderedProviders.map(MusicProvider::observePlaylists)) { playlists ->
            playlists
                .asList()
                .flatten()
                .sortedWith(
                    compareBy<PlaylistSummary>(
                        { it.provider.ordinal },
                        PlaylistSummary::title,
                    ),
                )
        }
    }

    override fun observePlaylist(ref: PlaylistRef): Flow<Playlist?> {
        return provider(ref.provider).observePlaylist(ref.id)
    }

    override fun observeSavedTracks(): Flow<SavedTracksResult> {
        val savedTracksProviders = orderedProviders.filter { it.id == MusicProviderId.YandexMusic }
        if (savedTracksProviders.isEmpty()) {
            return flowOf(SavedTracksResult.PrivateLibrary)
        }

        return combine(savedTracksProviders.map(MusicProvider::observeSavedTracks)) { results ->
            results.firstOrNull { it !is SavedTracksResult.PrivateLibrary } ?: SavedTracksResult.PrivateLibrary
        }
    }

    override suspend fun refreshAll() {
        coroutineScope {
            orderedProviders.map { provider ->
                async {
                    provider.refreshAvailability()
                    provider.refreshPlaylists()
                }
            }.awaitAll()
        }
    }

    override suspend fun refreshPlaylist(ref: PlaylistRef) {
        provider(ref.provider).refreshPlaylist(ref.id)
    }

    override suspend fun refreshSavedTracks() {
        coroutineScope {
            orderedProviders.map { provider ->
                async { provider.refreshSavedTracks() }
            }.awaitAll()
        }
    }

    private fun provider(id: MusicProviderId): MusicProvider {
        return checkNotNull(providersById[id]) {
            "Music provider $id is not registered."
        }
    }
}
